src/gather.jl

```diff
@@ -0,0 +1,49 @@
+export gather, gather_indices
+
+"""
+    gather!(dst, src, idx)
+
+Reverse operation of scatter, which gather data in `src` to destination according to `idx`.
+For each index `k` in `idx`, assign values to `dst` according to
+
+    dst[k...] = src[idx[k]...]
+
+# Arguments
+- `dst`: the destination where data would be assigned to.
+- `src`: the source data to be assigned.
+- `idx`: the mapping for assignment from source to destination.
+"""
+function gather!(dst::AbstractArray{T,N}, src::AbstractArray{T}, idx::AbstractArray{<:IntOrTuple,N}) where {T,N}
+    @assert size(dst) == size(idx) "dst and idx must have the same size."
+    for k = CartesianIndices(idx)
+        @inbounds dst[k] = src[idx[k]]
```

Comment:
Before `@inbounds`, should we check? Possibly like this you can still say `@inbounds gather(...` to disable it:
```
@boundscheck _checkin(idx, src)

_checkin(idx::AbstractArray{<:Integer}, arr) = 
    minimum(idx) in LinearIndices(arr) && maximum(idx) in LinearIndices(arr) || throw("some bounds error")
    
function _checkin(idx::AbstractArray{Tuple}, arr)
    for d in 1:ndims(arr)
        maximum(I->I[d], idx) in axes(arr,d) || error()
    end
end 
``` 
Surely there's a nicer way of treating the tuple case than my sketch here.

Reply:
Check new commits please.